source/OptimaJet.DWKit.StarterApplication/Services/Workflow/WorkflowProductService.cs

```diff
@@ -133,59 +124,126 @@ public async Task ProductProcessChangedAsync(ProductProcessChangedArgs args)
                  .ThenInclude(pr => pr.Owner)
                  .FirstOrDefaultAsync();
+
             if (product == null)
             {
                 Log.Error($"Could find Product for ProcessId={args.ProcessId}");
                 return;
             }
 
+            var tasks = await ReassignUserTasksForProduct(product, args);
+
+            // Clear the WorkflowComment
+            if (!String.IsNullOrWhiteSpace(product.WorkflowComment))
+            {
+                // Clear the comment
+                product.WorkflowComment = "";
+                await ProductRepository.UpdateAsync(product);
+            }
+        }
+
+        public async Task<List<UserTask>> ReassignUserTasksForProduct(Product product, ProductProcessChangedArgs args = null)
+        {
+            var instance = await ProductWorkflowRepository.GetAsync(product.Id);
+
+            var comment = GetCurrentTaskComment(product);
+
             await RemoveTasksByProductId(product.Id);
 
             // Find all users who could perform the current activity and create tasks for them
-            var workflowUserIds = Runtime.GetAllActorsForDirectCommandTransitions(args.ProcessId, activityName: args.CurrentActivityName).ToList();
-            var users = UserRepository.Get().Where(u => workflowUserIds.Contains(u.WorkflowUserId.GetValueOrDefault().ToString())).ToList();
-            var workflowComment = product.WorkflowComment;
+            var workflowUserIds = Runtime
+                .GetAllActorsForDirectCommandTransitions(product.Id, activityName : instance.ActivityName)
+                .ToList();
+
+            var users = UserRepository.Get()
+                .Where(u => workflowUserIds.Contains(u.WorkflowUserId.GetValueOrDefault().ToString()))
+                .ToList();
+
+            List<UserTask> result = new List<UserTask>();
+
             foreach (var user in users)
             {
-                var task = new UserTask
+                var userTask = new UserTask
                 {
                     UserId = user.Id,
                     ProductId = product.Id,
-                    ActivityName = args.CurrentActivityName,
-                    Status = args.CurrentState,
-                    Comment = workflowComment
+                    ActivityName = instance.ActivityName,
+                    Status = instance.StateName,
+                    Comment = comment
                 };
-                task = await TaskRepository.CreateAsync(task);
 
-                var messageParms = new Dictionary<string, object>()
+                var createdUserTask = await TaskRepository.CreateAsync(userTask);
+
+                await SendNotificationForTask(userTask, user, product, args);
+
+                result.Add(createdUserTask);
+            }
+
+            return result;
+        }
+
+        private async Task SendNotificationForTask (UserTask task, User user, Product product, ProductProcessChangedArgs args = null)
```

Comment:
passings `args` through was initially/partially the reason I didn't more cleanly refactor things at first.

Reply:
I don't know why `args` is a parameter.
* you can remove it from the Log statement (or move the Log statement to `ProductProcessChangedAsync` where it is more relevant).
* fromActivity is not used in the notification so that entry in the messageParms can be dropped.

Reply:
args is a param because it contained information that isn't used in the tasks or notifications
 - ExecutingCommand
 - Used on creation, but not re-assignment (due to there just not being a ProcessChange (from DWKit)
   - PreviousActivityName
   - CurrentState

fromActivity _IS_ used in the notification, in the middle of the Dictionary

> moving log to ProductProcessChangedAsync
 - we'd get one log for the whole thing, instead of per-task, but at this point it feels like bike-shedding, / also I think I'd prefer to just remove the log, and pass PreviousActivityName for the notification explicitly.